File: scripts/fetch_proposals.py
"""Fetch Snapshot proposals for a space and save as JSON fixtures.

These fixtures are the ground truth everything downstream uses:
  - LLM extraction prototyping  (scripts/extract.py, later)
  - Policy engine unit tests    (app/policy/tests/)
  - Decision-card calibration   (pre-digested onboarding)
  - Regression harness          (tests/integration/fixtures/proposals/)

Usage:
    python scripts/fetch_proposals.py
    python scripts/fetch_proposals.py --space arbitrumfoundation.eth --limit 50
    python scripts/fetch_proposals.py --state closed --limit 100
    python scripts/fetch_proposals.py --out data/proposals-snap-$(date +%F)
"""

from __future__ import annotations

import argparse
import json
import sys
from collections import Counter
from pathlib import Path
from typing import Any

import httpx

SNAPSHOT_HUB = "https://hub.snapshot.org/graphql"
BATCH_SIZE = 20
TIMEOUT_S = 30.0

PROPOSALS_QUERY = """
query Proposals($space: String!, $first: Int!, $skip: Int!) {
  proposals(
    first: $first,
    skip: $skip,
    where: { space: $space },
    orderBy: "created",
    orderDirection: desc
  ) {
    id
    title
    body
    choices
    start
    end
    snapshot
    state
    author
    created
    type
    scores
    scores_total
    votes
    quorum
    discussion
    link
    ipfs
    space { id name }
  }
}
"""


def fetch_proposals(space: str, limit: int) -> list[dict[str, Any]]:
    """Page through Snapshot's GraphQL API until we have `limit` proposals
    or the space is exhausted. Newest first."""
    results: list[dict[str, Any]] = []
    skip = 0
    with httpx.Client(timeout=TIMEOUT_S, headers={"Content-Type": "application/json"}) as client:
        while len(results) < limit:
            want = min(BATCH_SIZE, limit - len(results))
            payload = {
                "query": PROPOSALS_QUERY,
                "variables": {"space": space, "first": want, "skip": skip},
            }
            resp = client.post(SNAPSHOT_HUB, json=payload)
            resp.raise_for_status()
            body = resp.json()
            if "errors" in body:
                raise RuntimeError(f"Snapshot GraphQL error: {body['errors']}")
            batch = body.get("data", {}).get("proposals") or []
            if not batch:
                break
            results.extend(batch)
            skip += len(batch)
            print(f"  fetched {len(results)}/{limit}...", file=sys.stderr)
            if len(batch) < want:
                break
    return results[:limit]


def save_proposals(proposals: list[dict[str, Any]], out_dir: Path) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)
    index: list[dict[str, Any]] = []
    for p in proposals:
        pid = p["id"]
        path = out_dir / f"{pid}.json"
        path.write_text(json.dumps(p, indent=2, sort_keys=True) + "\n")
        index.append(
            {
                "id": pid,
                "title": p.get("title"),
                "state": p.get("state"),
                "created": p.get("created"),
                "end": p.get("end"),
                "votes": p.get("votes"),
                "author": p.get("author"),
                "type": p.get("type"),
            }
        )
    (out_dir / "index.json").write_text(json.dumps(index, indent=2) + "\n")


def summarize(proposals: list[dict[str, Any]]) -> None:
    states = Counter(p.get("state") for p in proposals)
    types = Counter(p.get("type") for p in proposals)
    body_lens = [len(p.get("body") or "") for p in proposals]
    avg_body = sum(body_lens) // max(len(body_lens), 1)
    print("", file=sys.stderr)
    print(f"Total:     {len(proposals)}", file=sys.stderr)
    print(f"By state:  {dict(states)}", file=sys.stderr)
    print(f"By type:   {dict(types)}", file=sys.stderr)
    print(f"Body avg:  {avg_body} chars (min {min(body_lens, default=0)}, max {max(body_lens, default=0)})", file=sys.stderr)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0] if __doc__ else "")
    ap.add_argument("--space", default="arbitrumfoundation.eth", help="Snapshot space (default: arbitrumfoundation.eth)")
    ap.add_argument("--limit", type=int, default=30, help="Max proposals to fetch (default: 30)")
    ap.add_argument("--out", type=Path, default=Path("data/proposals"), help="Output directory")
    ap.add_argument(
        "--state",
        choices=["active", "closed", "pending", "all"],
        default="all",
        help="Filter by state AFTER fetching (default: all)",
    )
    args = ap.parse_args()

    print(f"Fetching up to {args.limit} proposals from {args.space}...", file=sys.stderr)
    proposals = fetch_proposals(args.space, args.limit)

    if args.state != "all":
        before = len(proposals)
        proposals = [p for p in proposals if p.get("state") == args.state]
        print(f"Filtered state={args.state}: {before} -> {len(proposals)}", file=sys.stderr)

    save_proposals(proposals, args.out)
    print(f"Wrote {len(proposals)} proposals to {args.out}/", file=sys.stderr)
    summarize(proposals)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/fetch-proposals.ts
/**
 * Fetch Snapshot proposals for a space and save as JSON fixtures.
 *
 * Usage:
 *   npm run fetch:proposals
 *   npm run fetch:proposals -- --space arbitrumfoundation.eth --limit 50
 *   npm run fetch:proposals -- --state closed --limit 100
 */

import { writeFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { parseArgs } from 'node:util';

const SNAPSHOT_HUB = 'https://hub.snapshot.org/graphql';
const BATCH_SIZE = 20;

const QUERY = `
query Proposals($space: String!, $first: Int!, $skip: Int!) {
  proposals(
    first: $first,
    skip: $skip,
    where: { space: $space },
    orderBy: "created",
    orderDirection: desc
  ) {
    id
    title
    body
    choices
    start
    end
    snapshot
    state
    author
    created
    type
    scores
    scores_total
    votes
    quorum
    discussion
    link
    ipfs
    space { id name }
  }
}`;

type Proposal = {
  id: string;
  title?: string;
  body?: string;
  state?: string;
  created?: number;
  end?: number;
  author?: string;
  type?: string;
  votes?: number;
};

async function fetchProposals(space: string, limit: number): Promise<Proposal[]> {
  const results: Proposal[] = [];
  let skip = 0;
  while (results.length < limit) {
    const want = Math.min(BATCH_SIZE, limit - results.length);
    const res = await fetch(SNAPSHOT_HUB, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        query: QUERY,
        variables: { space, first: want, skip },
      }),
    });
    if (!res.ok) {
      throw new Error(`Snapshot HTTP ${res.status}: ${await res.text()}`);
    }
    const body = (await res.json()) as { data?: { proposals?: Proposal[] }; errors?: unknown };
    if (body.errors) {
      throw new Error(`Snapshot GraphQL error: ${JSON.stringify(body.errors)}`);
    }
    const batch = body.data?.proposals ?? [];
    if (batch.length === 0) break;
    results.push(...batch);
    skip += batch.length;
    console.error(`  fetched ${results.length}/${limit}...`);
    if (batch.length < want) break;
  }
  return results.slice(0, limit);
}

async function main() {
  const { values } = parseArgs({
    options: {
      space: { type: 'string', default: 'arbitrumfoundation.eth' },
      limit: { type: 'string', default: '30' },
      out: { type: 'string', default: 'data/proposals' },
      state: { type: 'string', default: 'all' },
    },
  });

  const space = values.space!;
  const limit = Number(values.limit);
  const outDir = values.out!;
  const state = values.state!;

  console.error(`Fetching up to ${limit} proposals from ${space}...`);
  let proposals = await fetchProposals(space, limit);

  if (state !== 'all') {
    const before = proposals.length;
    proposals = proposals.filter((p) => p.state === state);
    console.error(`Filtered state=${state}: ${before} -> ${proposals.length}`);
  }

  mkdirSync(outDir, { recursive: true });
  const index = proposals.map((p) => ({
    id: p.id,
    title: p.title,
    state: p.state,
    created: p.created,
    end: p.end,
    votes: p.votes,
    author: p.author,
    type: p.type,
  }));

  for (const p of proposals) {
    writeFileSync(join(outDir, `${p.id}.json`), JSON.stringify(p, null, 2) + '\n');
  }
  writeFileSync(join(outDir, 'index.json'), JSON.stringify(index, null, 2) + '\n');
  console.error(`Wrote ${proposals.length} proposals to ${outDir}/`);

  // Summary
  const states: Record<string, number> = {};
  const types: Record<string, number> = {};
  const bodyLens: number[] = [];
  for (const p of proposals) {
    const s = p.state ?? 'unknown';
    const t = p.type ?? 'unknown';
    states[s] = (states[s] ?? 0) + 1;
    types[t] = (types[t] ?? 0) + 1;
    bodyLens.push((p.body ?? '').length);
  }

  console.error('');
  console.error(`Total:     ${proposals.length}`);
  console.error(`By state:  ${JSON.stringify(states)}`);
  console.error(`By type:   ${JSON.stringify(types)}`);
  if (bodyLens.length > 0) {
    const total = bodyLens.reduce((a, b) => a + b, 0);
    const avg = Math.floor(total / bodyLens.length);
    console.error(`Body avg:  ${avg} chars (min ${Math.min(...bodyLens)}, max ${Math.max(...bodyLens)})`);
  }
}

main().catch((err) => {
  console.error(`ERROR: ${err instanceof Error ? err.message : String(err)}`);
  process.exit(1);
});
